### PR TITLE
[IMPROVE] User API Redesign

### PR DESCRIPTION
**Pull request by AlexanderKanakis**

### CODE REVIEW

**Comment by ear-dev** [comment]
@bhardwajaditya can you review this PR please.  It fixes the v1/users endpoint to conform to openAPI specs.  The spec itself is here: https://github.com/WideChat/Rocket.Chat.openApi/blob/master/usersApi.yaml

FYI - Alexander is cleaning up the api spec file to represent only just the changes to the users endpoint for GET, POST, PUT and DELETE.   We will not be implementing any changes to the current way /users works.... i.e. users.info, etc....  We'll just leave all that in because we can do everything we need with the changes to just v1/users and v1/users/{id}

**Comment by ear-dev** [comment]
@Shailesh351 Aditya is feeling sick again and will likely be out for a couple of days....  can you take a look at this PR please?  Thanks.

**Comment by Shailesh351** [comment]
Sure, Will do @ear-dev!

File: app/api/server/v1/users.js [hunk 1]
```diff
@@ -28,6 +28,229 @@ import { resetTOTP } from '../../../2fa/server/functions/resetTOTP';
...
+
+		Meteor.runAsUser(this.userId, () => {
+			Meteor.call('deleteUser', userId, confirmRelinquish);
+		});
+
+		return API.v1.success();
+	},
+	put() {
+		check(this.bodyParams, {
+			data: Match.ObjectIncluding({
```
Comment (Shailesh351) [inline app/api/server/v1/users.js:211]: @ear-dev @AlexanderKanakis We have added `status` in `users.create` and `users.register`. Should we also have it for updating users(PUT & `users.update`)?

File: app/api/server/v1/users.js [hunk 2]
```diff
@@ -173,7 +396,7 @@ API.v1.addRoute('users.deactivateIdle', { authRequired: true }, {
 
 API.v1.addRoute('users.getPresence', { authRequired: true }, {
 	get() {
-		if (this.isUserFromParams()) {
+		if (this.userId === this.urlParams.userId) {
```
Comment (Shailesh351) [inline app/api/server/v1/users.js]: Is this change required?

File: tests/end-to-end/api/00-miscellaneous.js [hunk 3]
```diff
@@ -228,7 +228,6 @@ describe('miscellaneous', function() {
 					expect(res.body).to.have.property('offset');
 					expect(res.body).to.have.property('total');
 					expect(res.body).to.have.property('count');
-					expect(res.body.result[0]).to.have.property('_id');
```
Comment (Shailesh351) [inline tests/end-to-end/api/00-miscellaneous.js]: I'm not sure but are our changes affecting these test cases? I thought we just added additional endpoints.

**Review by Shailesh351 — COMMENTED** [review]
I have tested it locally and it works perfectly!

While using API specs, I was not able to directly run the endpoint from spotlight studio.
@AlexanderKanakis Can you please also add headers in the spec which are required for auth?

And for some responses i.e `/users` It shows response as an array of users`[{user:{}}]` but actually, it's JSON object with `{users:[{},{}], count, offset, total, success}`. Can you please update that if required?

File: tests/end-to-end/api/00-miscellaneous.js (hunk 3 above)
Comment (AlexanderKanakis) [reply tests/end-to-end/api/00-miscellaneous.js]: These lines are restored.

File: app/api/server/v1/users.js (hunk 2 above)
Comment (AlexanderKanakis) [reply app/api/server/v1/users.js]: Reverted change.

File: app/api/server/v1/users.js (hunk 1 above)
Comment (AlexanderKanakis) [reply app/api/server/v1/users.js:211]: Added `status` in PUT. We do not intent to use `users.update`.

**Comment by ear-dev** [comment]
@AlexanderKanakis we will merge this now, but please update the openApi spec as per Shailesh recommendation.  Thanks!

**Review by Shailesh351 — APPROVED** [review]
